relay-sampling/src/evaluation.rs

```diff
@@ -24,26 +30,146 @@ fn pseudo_random_from_uuid(id: Uuid) -> f64 {
     generator.sample(dist)
 }
 
+/// The amount of matches for each reservoir rule in a given project.
+pub type ReservoirCounters = Arc<Mutex<BTreeMap<RuleId, i64>>>;
+
+/// Utility for evaluating reservoir-based sampling rules.
+///
+/// A "reservoir limit" rule samples every match until its limit is reached, after which
+/// the rule is disabled.
+///
+/// This utility uses a dual-counter system for enforcing this limit:
+///
+/// - Local Counter: Each relay instance maintains a local counter to track sampled events.
+///
+/// - Redis Counter: For processing relays, a Redis-based counter provides synchronization
+///   across multiple relay-instances. When incremented, the Redis counter returns the current global
+///   count for the given rule, which is then used to update the local counter.
+#[derive(Debug)]
+pub struct ReservoirEvaluator {
+    counters: ReservoirCounters,
+    #[cfg(feature = "redis")]
+    redis_pool: Option<Arc<RedisPool>>,
+    #[cfg(feature = "redis")]
+    org_id: Option<u64>,
+}
+
+impl ReservoirEvaluator {
+    /// Constructor for [`ReservoirEvaluator`].
+    pub fn new(counters: ReservoirCounters) -> Self {
+        Self {
+            counters,
+            #[cfg(feature = "redis")]
+            org_id: None,
+            #[cfg(feature = "redis")]
+            redis_pool: None,
+        }
+    }
+
+    /// Sets the Redis pool and organiation ID for the [`ReservoirEvaluator`].
+    ///
+    /// These values are needed to synchronize with Redis.
+    #[cfg(feature = "redis")]
+    pub fn set_redis(mut self, org_id: u64, redis_pool: Arc<RedisPool>) -> Self {
+        self.org_id = Some(org_id);
+        self.redis_pool = Some(redis_pool);
+        self
+    }
+
+    #[allow(unused_variables)]
+    fn redis_count(&self, rule: RuleId, rule_expiry: Option<&DateTime<Utc>>) -> Option<i64> {
+        #[cfg(feature = "redis")]
+        if let (Some(pool), Some(org_id)) = (self.redis_pool.as_ref(), self.org_id) {
+            let key = ReservoirRuleKey::new(org_id, rule);
+
+            let mut redis_client = pool.client().ok()?;
+            let mut redis_connection = redis_client.connection().ok()?;
+
+            if redis_sampling::set_redis_expiry(&mut redis_connection, &key, rule_expiry).is_err() {
+                relay_log::error!("failed to set redis reservoir rule expiry");
+            }
```

Comment:
@jan-auer  I wasn't sure what we would do if this returned an error, so I opted for just setting the expiry every time as we mentioned, then we could think about optimizing later.

Reply:
We could log and ignore the error on EXPIRE like you do here, and handle the increment error gracefully. The more correct version, however, would be: Encapsulate the two calls into a single function that throws an error, and if the error occurs simply do _not_ match the rule. 

That would mean if Redis is unavailable we skip all reservoir rules and apply other matching rules. A graceful and predictable fallback behavior that does not cause excess indexing.

Reply:
ok, theres an edge case if incrementing works but setting redis expiry doesn't in that case, but that should be very unlikely. If that happens htough, we'll count up without sampling.